specutils/fitting/fitmodels.py

```diff
@@ -1,20 +1,246 @@
+        line type (``line_type``) and index of line center (``line_center_index``) 
+        are stored for each line.
+    """
+
+    # Threshold based on noise estimate and factor.
+    uncertainty = spectrum.uncertainty
+    inds = np.where(np.abs(spectrum.flux) > (noise_factor*uncertainty.array)*spectrum.flux.unit)[0]
+    pos_inds = inds[spectrum.flux.value[inds] > 0]
+    line_inds_grouped = _consecutive(pos_inds, stepsize=1)
+
+    if len(line_inds_grouped[0]) > 0:
+        emission_inds = [inds[np.argmax(spectrum.flux.value[inds])] for inds in line_inds_grouped]
+    else:
+        emission_inds = []
+
+    #
+    # Find the absorption lines
+    #
+
+    neg_inds = inds[spectrum.flux.value[inds] < 0]
+    line_inds_grouped = _consecutive(neg_inds, stepsize=1)
+
+    if len(line_inds_grouped[0]) > 0:
+        absorption_inds = [inds[np.argmin(spectrum.flux.value[inds])] for inds in line_inds_grouped]
+    else:
+        absorption_inds = []
+
+    #
+    # Create the QTable to return the lines
+    #
+
+    qtable = QTable()
+    qtable['line_center'] = list(itertools.chain(*[spectrum.spectral_axis.value[emission_inds],
+                                             spectrum.spectral_axis.value[absorption_inds]]))*spectrum.spectral_axis.unit
+    qtable['line_type'] = ['emission']*len(emission_inds) + ['absorption']*len(absorption_inds)
+    qtable['line_center_index'] = list(itertools.chain(*[emission_inds, absorption_inds]))
+
+    return qtable
+
+
+def find_lines_derivative(spectrum, flux_threshold):
+    """
+    Find the emission and absorption lines in a spectrum. The method
+    here is a based on finding the zero crossings in the derivative 
+    of the spectrum.
+
+    Parameters
+    ----------
+    spectrum : Spectrum1D
+        The spectrum object over which the equivalent width will be calculated.
+
+    flux_threshold : float or `~astropy.units.Quantity`
+        Threshold on line finding.
+
+    Returns
+    -------
+    qtable: `~astropy.table.QTable`
+        Table of emission and absorption lines. Line center (``line_center``),
+        line type (``line_type``) and index of line center (``line_center_index``) 
+        are stored for each line.
+    """
+
+    # Take the derivative to find the zero crossings which correspond to
+    # the peaks (positive or negative)
+    kernel = [1, 0, -1]
+    dY = convolve(spectrum.flux, kernel, 'valid')
+
+    # Use sign flipping to determine direction of change
+    S = np.sign(dY)
+    ddS = convolve(S, kernel, 'valid')
+
+    # Add units if needed.
+    if isinstance(flux_threshold, (int, float)):
+        flux_threshold = float(flux_threshold) * spectrum.flux.unit
+
+    #
+    # Emmision lines
+    #
+
+    # Find all the indices that appear to be part of a +ve peak
+    candidates = np.where(dY > 0)[0] + (len(kernel) - 1)
+    line_inds = sorted(set(candidates).intersection(np.where(ddS == -2)[0] + 1))
```

Comment:
One other thing I just noticed: shouldn't these all be numpy set operations instead of python operations?

Reply:
Ah, numpy has set operations?

Reply:
Yep, they're quite close to this but return arrays instead of set objects.  Can do this later though.